File: algorithm-design/hackerrank/icecream_parlor/icecream_parlor.py
#!/usr/bin/env python3
# https://www.hackerrank.com/challenges/icecream-parlor

import os
import unittest

from typing import List


def icecream_parlor(m: int, arr: List[int]) -> List[int]:
    n = len(arr)
    mapped = {}
    for i in range(0, n):
        mapped[arr[i]] = i
    for i in range(0, n):
        first = i
        value = m - arr[i]
        if value in mapped:
            second = mapped[value]
            if first == second:
                continue
            return [first + 1, second + 1]
    return [0, 0]


class TestCode(unittest.TestCase):
    def runner(self, name):
        path = os.path.join(os.path.split(__file__)[0], f'input{name}.txt')
        with open(path, 'r') as handle:
            lines = handle.readlines()
            lines = [line.strip() for line in lines]
        count = int(lines[0])
        results = [None] * count
        for i in range(count):
            m = int(lines[1 + i * 3])
            arr = [int(item) for item in lines[3 + i * 3].split(' ')]
            results[i] = icecream_parlor(m, arr)
        path = os.path.join(os.path.split(__file__)[0], f'output{name}.txt')
        with open(path, 'r') as handle:
            lines = handle.readlines()
        expected = [[int(item) for item in line.split(' ')] for line in lines]
        self.assertEqual(expected, results)

    def test_example(self):
        self.runner('_example')

    def test_02(self):
        self.runner('02')

    def test_same(self):
        expected = [3, 4]
        result = icecream_parlor(6, [3, 1, 2, 4])
        self.assertEqual(expected, result)

    def test_none(self):
        expected = [0, 0]
        result = icecream_parlor(10, [3, 1, 2, 4])
        self.assertEqual(expected, result)

File: algorithms/code/hackerrank/icecream_parlor/icecream_parlor.rs
// https://www.hackerrank.com/challenges/icecream-parlor

use crate::hackerrank::tester::{read_input, write_and_check_output};
use std::collections::HashMap;

pub fn icecream_parlor(m: i32, arr: &[i32]) -> Vec<i32> {
    let n = arr.len() as i32;
    let mut mapped = HashMap::<i32, i32>::new();
    for i in 0..n {
        mapped.insert(arr[i as usize], i);
    }
    for i in 0..n {
        let first = i;
        let value = m - arr[i as usize];
        if let Some(second) = mapped.get(&value) {
            if first == *second {
                continue;
            }
            return vec![first + 1, second + 1];
        }
    }
    vec![0, 0]
}

pub fn main_tested(name: &str) {
    let origin = file!();
    let input = read_input(origin, name);
    let mut results: Vec<String> = Vec::new();
    let t = input[0][0].parse::<u32>().unwrap();
    let mut offset = 1;
    for _ in 0..t {
        let m = input[offset][0].parse::<i32>().unwrap();
        let arr: Vec<i32> = input[offset + 2]
            .iter()
            .map(|s| s.parse::<i32>().unwrap())
            .collect();
        offset += 3;
        results.push(
            icecream_parlor(m, &arr)
                .iter()
                .map(|i| i.to_string())
                .collect::<Vec<String>>()
                .join(" "),
        );
    }
    write_and_check_output(origin, name, &results);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        main_tested("_example");
    }

    #[test]
    fn test02() {
        main_tested("02");
    }

    #[test]
    fn test_same() {
        let expected = vec![3, 4];
        let result = icecream_parlor(6, &[3, 1, 2, 4]);
        assert_eq!(expected, result);
    }

    #[test]
    fn test_none() {
        let expected = vec![0, 0];
        let result = icecream_parlor(10, &[3, 1, 2, 4]);
        assert_eq!(expected, result);
    }
}
